README.md
# Py-Pag

O py-pag é uma pequena lib construída para facilitar a integração de assinaturas manuais no [PagSeguro](pagseguro.uol.com.br).

O Py-Pag depende:

* arrow
* BeautifulSoup4
* requests
* Unidecode

> Todos instalados automaticamente.


## Instalando

Use o seguinte comando para instalar:

`pip install py_pag`


## Começando

Abaixo a descrição de algumas classes e suas funções


### Transação simples

Iniciando uma compra de produto pontual

```python
from py_pag import PagSeguroTransaction

# Abaixo crie um objecto de transação
transaction = PagSeguroTransaction('sua_token', 'seu_email', reference='Referencia de seu sistema')

# Crie também o item a ser cobrado, um simples dicionário
item = {
    'id': 'id',  # Nome do item no pagseguro
    'description': 'item description',  # Descrição do item
    'quantity': 1,  # Quantidade de itens vendidos
    'amount': 3.14,  # Valor do item
}

# Define qual o item da transação
transaction.set_item(item)

# Url para você redirecionar seu comprador para o PagSeguro
url = transaction.get_checkout_url()
```


### Criando uma assinatura

Para iniciar uma assinatura no pagseguro use:


```python
from py_pag import PagSeguroSignature

# Abaixo crie um objecto de transação
transaction = PagSeguroSignature('sua_token', 'seu_email', 'nome da asisnatura', 'descricao', price=3.14, reference='Referencia de seu sistema')

# Caso sua assinatura não seja mensal passe também o argumento period com uma das opções:
# WEEKLY, BIMONTHLY, TRIMONTHLY, SEMIANNUALLY, YEARLY

# Url para você redirecionar seu comprador para o PagSeguro
url = transaction.get_checkout_url()
```


### Recebendo notificaçoes

Após qualquer uma das ações acima o pagseguro irá processar e lhe enviar(POST) uma notificação.
A notificação do pagseguro é um post contendo `notificationCode` e `notificationType`.

O `notificationType` pode ser Transaction ou Preapproval, para vendas simples e assinaturas respectivamente.


##### Notificação de transição

Caso seja uma Transaction use:

```python
from py_pag import PagSeguroNotificationHandler

# Instancie o handler passando o código recebido pelo pagseguro.
handler = PagSeguroNotificationHandler('sua_token', 'seu_email', 'notificationCode')

# Use o get_notification_response para obter as informações refentes a essa transação
response = handler.get_notification_response()

# Status da transação
response.status  # 'Paga'
```


##### Notificação de assinatura

Caso seja uma assinatura use:

```python
from py_pag import PagSeguroNotificationSignatureHandler

# Instancie o handler passando o código recebido pelo pagseguro.
handler = PagSeguroNotificationSignatureHandler('sua_token', 'seu_email', 'notificationCode')

# Use o get_notification_response para obter as informações refentes a essa transação
response = handler.get_notification_response()

if response.status.lower() == 'active':  # 3 estados possíveis: INITIATED, PENDING e ACTIVE
    # Faça a cobrança, veja abaixo usando o response.code
    # Para mais de uma cobrança salve o response.code em seu banco de dados
```


### Efetuando cobranças

Uma vez que sua assinatura já está ativa podemos começar a fazer as cobranças de acordo com a periodicidade acordada:


```python
from py_pag import PagSeguroSignatureCharger

# Itens a serem cobrados
items = [{
    'id': 'nome',  # Nome do item no pagseguro
    'description': 'item description',  # Descrição do item
    'quantity': 1,  # Quantidade de itens vendidos
    'amount': 3.14,  # Valor do item
}]

# Atenção que o valor cobrado DEVE ser exatamento o mesmo da criação da assinatura
charger = PagSeguroSignatureCharger('sua_token', 'seu_email', 'codigo_da_assinatura', items)

# O charge() irá cobrar ou levantar uma exception com a mensagem do pagseguro
# O retorno consiste do códgio da operação e data da mesma.
# Após uma cobrança bem sucedida o pagseguro irá lhe enviar uma nova notificação indicando que algo foi pago
# Esta notificação será uma transação normal.
codigo_da_operacao, data = charger.charge()
```


## Executando os testes

* Clone o repositório.
* Entre na pasta raiz e instale as dependências de dev: `pip install -r requirements-dev.txt`.
* Para rodar os testes: `py.test`.
* Para obter o coverage: `py.test --cov`


## Implementações futuras

* Alterar PagSeguroTransaction para trabalhar com multiplos produtos.
* Criar documentação sobre url_redirect e outras opções.


CHANGES.md
### 0.1.1

(21 de Agosto de 2015)

- Corrige verificação de erros do pagseguro


tests/utils.py
# encoding: utf-8

def post_fake(*args, **kwargs):
    class ResponseFake(object):
        content = """
        <?xml version="1.0" encoding="ISO-8859-1"?>  
        <checkout>  
            <code>8CF4BE7DCECEF0F004A6DFA0A8243412</code>  
            <date>2010-12-02T10:11:28.000-02:00</date>  
        </checkout>
        """
        status_code = 200
    return ResponseFake()

def post_fake_without_code(*args, **kwargs):
    class ResponseFake(object):
        content = """
        <?xml version="1.0" encoding="ISO-8859-1"?>  
        <checkout>  
            <date>2010-12-02T10:11:28.000-02:00</date>  
        </checkout>
        """
        status_code = 200
    return ResponseFake()

def post_fake_fail(*args, **kwargs):
    class ResponseFake(object):
        content = """
        <?xml version="1.0" encoding="ISO-8859-1"?>
        <errors>
            <error>
                <code>11004</code>
                <message>Currency is required.</message>
            </error>
            <error>
                <code>11005</code>
                <message>Currency invalid value: 100</message>
            </error>
        </errors>
        """
        status_code = 400
    return ResponseFake()

def post_fake_unauthorized(*args, **kwargs):
    class ResponseFake(object):
        content = """
        <p>Unauthorized</p>
        """
        status_code = 401
    return ResponseFake()

def get_fake_notification_not_found(*args, **kwargs):
    class ResponseFake(object):
        content = "<p>Not Found</p>"
        status_code = 404
    return ResponseFake()

def get_fake_notification(*args, **kwargs):
    class ResponseFake(object):
        content = """
        <!--?xml version="1.0" encoding="ISO-8859-1" standalone="yes"?-->
        <transaction>
            <date>2011-02-10T16:13:41.000-03:00</date>
            <code>9E884542-81B3-4419-9A75-BCC6FB495EF1</code>
            <reference>REF1234</reference>
            <type>1</type>
            <status>3</status>
            <paymentmethod>
                <type>1</type>
                <code>101</code>
            </paymentmethod>
            <grossamount>49900.00</grossamount>
            <discountamount>0.00<discountamount>
            <creditorfees>
                <intermediationrateamount>0.40</intermediationrateamount>
                <intermediationfeeamount>1644.80</intermediationfeeamount>
            </creditorfees>
            <netamount>49900.00</netamount>
            <extraamount>0.00</extraamount>
            <installmentcount>1</installmentcount>
            <itemcount>2</itemcount>
            <items>
                <item>
                    <id>0001</id>
                    <description>Notebook Prata</description>
                    <quantity>1</quantity>
                    <amount>24300.00</amount>
                </item>
                <item>
                    <id>0002</id>
                    <description>Notebook Rosa</description>
                    <quantity>1</quantity>
                    <amount>25600.00</amount>
                </item>
            </items>
            <sender>
                <name>José Comprador</name>
                <email>comprador@uol.com.br</email>
                <phone>
                    <areacode>11</areacode>
                    <number>56273440</number>
                </phone>
            </sender>
            <shipping>
                <address>
                    <street>Av. Brig. Faria Lima</street>
                    <number>1384</number>
                    <complement>5o andar</complement>
                    <district>Jardim Paulistano</district>
                    <postalcode>01452002</postalcode>
                    <city>Sao Paulo</city>
                    <state>SP</state>
                    <country>BRA</country>
                </address>
                <type>1</type>
                <cost>21.50</cost>
            </shipping>
        </discountamount></discountamount></transaction>
        """
        status_code = 200
    return ResponseFake()

def get_fake_signature_notification(*args, **kwargs):
    class ResponseFake(object):
        content = """
        <!--?xml version="1.0" encoding="ISO-8859-1" standalone="yes"?-->
        <preApproval>
            <name>Seguro contra roubo do Notebook Prata</name>
            <code>C08984179E9EDF3DD4023F87B71DE349</code>
            <date>2011-11-23T13:40:23.000-02:00</date>
            <tracker>538C53</tracker>
            <status>ACTIVE</status>
            <reference>REF1234</reference>
            <lastEventDate>2011-11-25T20:04:23.000-02:00</lastEventDate>
            <charge>auto</charge>
                <sender>
                <name>Comprador Istambul</name>
                <email>c@i.com</email>
                <phone>
                <areaCode>11</areaCode>
                <number>30389678</number>
                </phone>
                <address>
                <street>ALAMEDA ITU</street>
                <number>78</number>
                <complement>ap. 2601</complement>
                <district>Jardim Paulista</district>
                <city>SAO PAULO</city>
                <state>SP</state>
                <country>BRASIL</country>
                <postalCode>01421000</postalCode>
                </address>
                </sender>
        </preApproval>
        """
        status_code = 200
    return ResponseFake()


def get_fake_charger(*args, **kwargs):
    class ResponseFake(object):
        content = """
        <!--?xml version="1.0" encoding="ISO-8859-1" standalone="yes"?-->
        <result>
        <transactionCode>D9AD1EA3DEB544A6A413E33BD4822225</transactionCode>
        <date>2011-08-19T14:47:59.000-03:00</date>
        </result>
        """
        status_code = 200
    return ResponseFake()

def get_fake_charger_error(*args, **kwargs):
    class ResponseFake(object):
        content = """
        <errors>
            <error>
            <code>11004</code>
            <message>Currency is required.</message>
            </error>
            <error>
            <code>11005</code>
            <message>Currency invalid value: ValorCurrencyInvalido</message>
            </error>
        </errors>
        """
        status_code = 400
    return ResponseFake()


tests/base_test_case.py
# encoding: utf-8

import unittest

class BaseTestCase(unittest.TestCase):
    def setUp(self):
        self.item = {
            'id': 'id',
            'description': 'item description',
            'quantity': 1,
            'amount': 3.14,
        }


tests/test_pagseguro_sender.py
# encoding: utf-8

import mock

from py_pag.pagseguro import PagSeguroSender
from py_pag.exceptions import ApiErrorException, UnauthorizedException

from .base_test_case import BaseTestCase
from .utils import *


class PagSeguroSenderTestCase(BaseTestCase):
    def setUp(self):
        self.data = {
            'name': 'User Name',
            'ddd': '38',
            'phone': '22222222',
            'email': 'email@teste.com',
            'city': 'City Name',
            'uf': 'MG',
        }

    def test_get_dict(self):
        sender = PagSeguroSender(**self.data)
        expected = {
            'senderName': self.data['name'],
            'senderAreaCode': self.data['ddd'],
            'senderPhone': self.data['phone'],
            'senderEmail': self.data['email'],
            'senderAddressCity': self.data['city'],
            'senderAddressState': self.data['uf'],
            'senderAddressCountry': 'BRA',
        }
        self.assertEqual(expected, sender.get_dados())

    def test_with_invalid_name(self):
        self.data['name'] = 'this name is really really long even for crazy standads'
        with self.assertRaises(ValueError):
            sender = PagSeguroSender(**self.data)

    def test_with_invalid_ddd(self):
        self.data['ddd'] = '1235'
        with self.assertRaises(ValueError):
            sender = PagSeguroSender(**self.data)

    def test_with_invalid_phone(self):
        self.data['phone'] = '112331241231'
        with self.assertRaises(ValueError):
            sender = PagSeguroSender(**self.data)

    def test_with_invalid_city(self):
        self.data['city'] = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit amet.'
        with self.assertRaises(ValueError):
            sender = PagSeguroSender(**self.data)

    def test_with_invalid_uf(self):
        self.data['uf'] = 'VAC'
        with self.assertRaises(ValueError):
            sender = PagSeguroSender(**self.data)


py_pag/pagseguro.py
# encoding: utf-8

import arrow
import requests
import re

from uuid import uuid4
from bs4 import BeautifulSoup
from unidecode import unidecode

from .exceptions import ApiErrorException, UnauthorizedException, NotificationNotFoundException


class PagSeguroBase(object):
    def __init__(self, token, email):
        self.email = email
        self.token = token
        self.headers = {
            'Content-type': 'application/x-www-form-urlencoded',
            'charset': 'UTF-8'
        }


class PagSeguroTransaction(PagSeguroBase):
    URLS = {
        'request': 'https://ws.pagseguro.uol.com.br/v2/checkout',
        'request_redirect': 'https://pagseguro.uol.com.br/v2/checkout/payment.html?code={code}',
    }

    def __init__(self, token, email, reference=None, item=None):
        self.item = item
        self.reference = reference
        self.sender = None
        self.redirect_url = None
        self.discount = None

        if not reference:
            self.reference = str(uuid4())
        else:
            self.reference = reference

        super(PagSeguroTransaction, self).__init__(token, email)

    def set_item(self, item):
        self.item = item

    def get_reference(self):
        return self.reference

    def get_code(self, soup):
        try:
            code = str(soup.find('code').string)
        except AttributeError:
            raise ApiErrorException('Code Not Found')
        return code

    def set_discount(self, value):
        if type(value) is not float:
            raise ValueError('Price should be float')
        self.discount = value

    def get_checkout_url(self):
        response = requests.post(self.URLS['request'], data=self.get_dados(), headers=self.headers)
        if response.status_code == 401:
            raise UnauthorizedException(u'invalid Token ou E-mail')
        soup = BeautifulSoup(response.content, 'html.parser')
        if response.status_code == 400:
            code = str(soup.errors.error.code.string)
            message = str(soup.errors.error.message.string)
            raise ApiErrorException('{code} - {message}'.format(code=code, message=message))
        code = self.get_code(soup)
        return self.URLS['request_redirect'].format(code=code)

    def set_sender(self, email, name):
        if name:
            name = name[:49]
            name = re.sub(' +', ' ', name)
        self.sender = {'email': email, 'name': name}

    def get_dados(self):
        return self.__dict__()

    def set_redirect_url(self, url):
        self.redirect_url = url

    def __dict__(self):
        data = {
            'token': self.token,
            'email': self.email,
            'currency': 'BRL',
            'reference': self.reference,
        }

        if self.item:
            data['itemId1'] = self.item['id'],
            data['itemDescription1'] = self.item['description'],
            data['itemAmount1'] = "{0:.2f}".format(round(self.item['amount'], 2)),
            data['itemQuantity1'] = self.item['quantity'],
        if self.sender:
            data['senderEmail'] = self.sender['email'],
            data['senderName'] = self.sender['name'],
        if self.redirect_url:
            data['redirectURL'] = self.redirect_url
        if self.discount:
            data['extraAmount'] = "-{0:.2f}".format(self.discount, 2)
        return data


class PagSeguroSignature(PagSeguroTransaction):
    URLS = {
        'request': 'https://ws.pagseguro.uol.com.br/v2/pre-approvals/request',
        'request_redirect': 'https://pagseguro.uol.com.br/v2/pre-approvals/request.html?code={code}',
    }

    TIPOS_PERIODOS = ['WEEKLY', 'MONTHLY', 'BIMONTHLY', 'TRIMONTHLY', 'SEMIANNUALLY', 'YEARLY']

    def __init__(self, token, email, name, description, price, reference=None, period='MONTHLY'):
        if len(name) > 100:
            raise ValueError('Invalid name, too long')
        if len(description) > 255:
            raise ValueError('Invalid Description, too long')
        if type(price) is not float:
            raise ValueError('Price should be float')
        if period not in self.TIPOS_PERIODOS:
            raise ValueError('Invalid period')

        self.name = name
        self.description = description
        self.price = price
        self.period = period
        self._set_end_date()

        super(PagSeguroSignature, self).__init__(token, email, reference)

    def __dict__(self):
        data = {
            'token': self.token,
            'email': self.email,
            'preApprovalName': self.name,
            'preApprovalDetails': self.description,
            'preApprovalAmountPerPayment': "{0:.2f}".format(round(self.price, 2)),
            'preApprovalPeriod': self.period,
            'preApprovalMaxPaymentsPerPeriod': 1,
            'preApprovalMaxAmountPerPeriod': "{0:.2f}".format(round(self.price, 2)),
            'preApprovalMaxTotalAmount': "{0:.2f}".format(round(self.price * 12, 2)),
            'preApprovalFinalDate': self.end_date,
            'reference': self.reference
        }

        if self.sender:
            data.update({'senderEmail': self.sender['email'], 'senderName': self.sender['name']})
        if self.redirect_url:
            data['redirectURL'] = self.redirect_url
        return data

    def _set_end_date(self):
        now = arrow.now()
        next_year = now.replace(years=+1).datetime
        next_year_iso = next_year.isoformat()
        self.end_date = next_year_iso[:-13] + next_year_iso[-6:]


class PagSeguroSender(object):
    def __init__(self, name=None, ddd=None, phone=None, email=None, city=None, uf=None):
        if name and len(name) > 50:
            raise ValueError('Invalid Name')
        if ddd and len(ddd) > 2:
            raise ValueError('Invalid DDD')
        if phone and (len(phone) < 7 or len(phone) > 9):
            raise ValueError('Invalid phone')
        if city and len(city) > 60:
            raise ValueError('Invalid city name')
        if uf and len(uf) is not 2:
            raise ValueError('Invalid UF')

        self.name = name
        self.ddd = ddd
        self.phone = phone
        self.email = email
        self.city = city
        self.uf = uf
        self.country = 'BRA'

    def __dict__(self):
        return {
            'senderName': self.name,
            'senderAreaCode': self.ddd,
            'senderPhone': self.phone,
            'senderEmail': self.email,
            'senderAddressCity': self.city,
            'senderAddressState': self.uf,
            'senderAddressCountry': self.country
        }

    def get_dados(self):
        return self.__dict__()


class PagSeguroNotificationHandler(PagSeguroBase):
    URLS = {
        'check_notification': 'https://ws.pagseguro.uol.com.br/v2/transactions/notifications/{code}?email={email}&token={token}',
    }

    def __init__(self, token, email, code):
        if not code:
            raise ValueError("No code provided")
        self.code = code
        super(PagSeguroNotificationHandler, self).__init__(token, email)

    def _valid_notification(self, soup):
        if soup.p and str(soup.p.string) == 'Not Found':
            raise NotificationNotFoundException('Notification not found')

        if soup.find('error'):
            code = str(soup.errors.error.code.string)
            message = str(soup.errors.error.message.string)
            raise ApiErrorException('{code} - {message}'.format(code=code, message=message))

    def get_check_notification_url(self):
        return self.URLS['check_notification'].format(code=self.code, email=self.email, token=self.token)

    def _get_notification_info(self):
        notification_url = self.get_check_notification_url()
        response = requests.get(notification_url, headers=self.headers)
        self.notification_response = response.content
        return self.notification_response

    def get_notification_response(self):
        soup = BeautifulSoup(self._get_notification_info())
        self._valid_notification(soup)

        pagseguro_response = PagSeguroNotificationResponse(
            str(soup.code.string),
            str(soup.reference.string),
            str(soup.type.string),
            str(soup.status.string),
            str(soup.grossamount.string),
            soup.sender
        )
        return pagseguro_response


class PagSeguroNotificationSignatureHandler(PagSeguroNotificationHandler):
    URLS = {
        'check_notification': 'https://ws.pagseguro.uol.com.br/v2/pre-approvals/notifications/{code}?email={email}&token={token}',
        'check_assinatura': 'https://ws.pagseguro.uol.com.br/v2/pre-approvals/{code}?email={email}&token={token}',
        'payment': 'https://ws.pagseguro.uol.com.br/v2/pre-approvals/payment/',
    }

    def get_signature_notification_info(self):
        soup = BeautifulSoup(self._get_notification_info())
        self._valid_notification(soup)

        self.data = soup.find('preapproval')
        return self.get_signature_notification()

    def get_signature_notification(self):
        data = self.data
        signature_response = PagSeguroSignatureResponse(
            str(data.code.string),
            str(data.date.string),
            str(data.tracker.string),
            str(data.status.string),
            str(data.reference.string),
            str(data.lasteventdate.string),
            str(data.charge.string),
            data.sender
        )
        return signature_response

    def get_notification_response(self):
        return self.get_signature_notification_info()


class PagSeguroAbstractResponse(object):
    """ Abstract class for a PagSeguro Response """

    def __init__(self):
        raise NotImplementedError  # pragma: no cover

    def format_sender_data(self, sender_soup):
        sender = {}
        try:
            sender['name'] = unidecode(str(sender_soup.find('name').string)),
        except:
            sender['name'] = u'Problema em identificar o nome'

        sender['name'] = str(sender_soup.find('email').string),
        return sender


class PagSeguroSignatureResponse(PagSeguroAbstractResponse):
    def __init__(self, code, date, tracker, status, reference, lasteventdate, charge, soup_sender):
        self.code = code
        self.date = arrow.get(date).datetime
        self.tracker = tracker
        self.status = status
        self.reference = reference
        self.lasteventdate = lasteventdate
        self.charge = charge
        self.sender = self.format_sender_data(soup_sender)


class PagSeguroNotificationResponse(PagSeguroAbstractResponse):
    STATUS = {
        1: u'Aguardando pagamento',
        2: u'Em análise',
        3: u'Paga',
        4: u'Disponível',
        5: u'Em disputa',
        6: u'Devolvida',
        7: u'Cancelada',
    }

    TRANSACTION_TYPE = {
        1: 'Pagamento',
        11: 'Assinatura'
    }

    def __init__(self, code, reference, transactiontype, status, amount, soup_sender):
        self.code = code
        self.reference = reference
        self.transactiontype = self.TRANSACTION_TYPE[int(transactiontype)]
        self.status = self.STATUS[int(status)]
        self.amount = amount
        self.sender = self.format_sender_data(soup_sender)


class PagSeguroSignatureCharger(PagSeguroTransaction):
    URLS = {
        'payment': 'https://ws.pagseguro.uol.com.br/v2/pre-approvals/payment/'
    }

    def __init__(self, token, email, code, items=[], reference=None):
        super(PagSeguroSignatureCharger, self).__init__(token, email, reference)
        self.code = code
        self.items = items
        self.reference = reference

    def get_data(self):
        data = {
            'email': self.email,
            'token': self.token,
            'preApprovalCode': self.code
        }
        if self.reference:
            data['reference'] = self.reference

        data.update(self.get_items_data())
        return data

    def get_items_data(self):
        data = {}
        for index, item in enumerate(self.items):
            data['itemId{index}'.format(index=index + 1)] = item['id']
            data['itemDescription{index}'.format(index=index + 1)] = item['description']
            data['itemAmount{index}'.format(index=index + 1)] = "{:.2f}".format(item['amount'])
            data['itemQuantity{index}'.format(index=index + 1)] = item['quantity']
        return data

    def charge(self):
        response = requests.post(self.URLS['payment'], data=self.get_data(), headers=self.headers)
        if response.status_code == 401:
            raise UnauthorizedException(u'invalid Token ou E-mail')
        soup = BeautifulSoup(response.content)
        if response.status_code == 400:
            code = str(soup.errors.error.code.string)
            message = str(soup.errors.error.message.string)
            raise ApiErrorException('{code} - {message}'.format(code=code, message=message))

        return (soup.transactionCode, soup.date)
